### PR TITLE
Fix moderated session presence checking

### PR DESCRIPTION
**Pull request by rosstimothy**
Addresses all of the issues that were preventing presence checking during moderated sessions from working as described in
[#18092](https://github.com/gravitational/teleport/issues/18092#issuecomment-1540900859).

Closes #18092

### CODE REVIEW

File: api/types/session_tracker_test.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	require.NoError(t, err)
+
+	// Presence cannot be updated for a non-existent user
+	err = s.UpdatePresence("alpaca")
+	require.ErrorIs(t, err, trace.NotFound("participant alpaca not found"))
+
+	// Presence is only updated for the matching user
+	require.NoError(t, s.UpdatePresence("fish"))
+	participants := s.GetParticipants()
+	require.Equal(t, clock.Now().UTC(), participants[0].LastActive)
```
Comment (zmb3) [inline api/types/session_tracker_test.go:54]: How does this work? `UpdatePresence` uses `time.Now`, not the fake clock.

File: lib/client/session.go [hunk 2]
```diff
@@ -492,17 +492,41 @@ func (ns *NodeSession) updateTerminalSize(ctx context.Context, s *tracessh.Sessi
...
+// by trying to use it before it has been initialized.
+// In those cases output is written to the stdout that
+// is configured for tsh so that output is not lost entirely.
+type sessionWriter struct {
+	tshOut  io.Writer
+	session *tracessh.Session
+}
+
+func (s *sessionWriter) Write(p []byte) (int, error) {
+	if s.session.Stdout != nil {
```
Comment (zmb3) [inline lib/client/session.go:506]: When does `Stdout` get set? Does that happen in another goroutine?

**Review by jentfoo — APPROVED** [review]
Looks reasonable, let me know if you want help testing

**Comment by ravicious** [comment]
I'm almost finished with reviewing this, though I'm no expert on the code being changed here. I'll try to test this on Monday.

File: lib/web/apiserver_test.go [hunk 3]
```diff
@@ -8778,19 +8912,183 @@ func TestModeratedSession(t *testing.T) {
...
+	// for is not present in the command itself
+	_, err = io.WriteString(peerStream, "echo llxmx | sed 's/x/a/g'\r\n")
+	require.NoError(t, err)
+	require.NoError(t, waitForOutput(peerStream, "llama"))
+	require.NoError(t, waitForOutput(moderatorStream, "llama"))
+
+	// run the presence check a few times
+	for i := 0; i < 3; i++ {
+		s.clock.Advance(30 * time.Second)
+		require.NoError(t, waitForOutput(moderatorStream, "Teleport > Please tap your MFA key"))
```
Comment (ravicious) [inline lib/web/apiserver_test.go:9037]: I'm not familiar with Go APIs for reading streams, but is `waitForOutput` guaranteed to receive the expected string even if it's written to the stream between `s.clock.Advance` and `waitForOutput`?

**Review by ravicious — COMMENTED** [review]
I tested moderated file transfer with cluster-wide per-session MFA turned on as well as regular file transfer and both work fine after the changes. Moderator presence checks work as expected as well.

I'm yet to review all the tests though.

**Review by ravicious — APPROVED** [review]

File: lib/web/terminal.go [hunk 4]
```diff
@@ -1032,6 +946,18 @@ type WSStream struct {
...
 	decoder *encoding.Decoder
 
+	// resizeC a channel to forward resize events so that
+	// they happen out of band and don't block reads
+	resizeC chan Envelope
+	// fileTransferRequestC is a channel to facilitate requesting a file transfer
+	fileTransferRequestC chan Envelope
+	// fileTransferDecisionC is a channel to facilitate responding to a file transfer
+	// with an approval or denial
+	fileTransferDecisionC chan Envelope
```
Comment (jakule) [inline lib/web/terminal.go:956]: Can we not keep these fields here? WSStream is used by Terminal and Command Handlers. The second one doesn't need them. IMO we should try to find an abstraction that holds only the required fields in a struct.

File: lib/web/terminal.go [hunk 5]
```diff
@@ -1040,23 +966,27 @@ type WSStream struct {
 	mu sync.Mutex
 	// ws the connection to the UI
 	ws WSConn
+
+	// ctx is used to signal when the terminal session is closing.
+	ctx context.Context
+	// cancel is used to terminate the terminal session.
+	cancel context.CancelFunc
```
Comment (jakule) [inline lib/web/terminal.go:973]: https://go.dev/blog/context-and-structs

File: lib/web/terminal.go [hunk 6]
```diff
@@ -1068,6 +998,199 @@ func (t *WSStream) writeError(err error) error {
...
+				select {
+				case <-t.ctx.Done():
+					return
+				case t.rawC <- Envelope{Payload: msg}:
+					return
+				}
+			}
+
+			if ty != websocket.BinaryMessage {
+				t.log.Warnf("Expected binary message, got %v", ty)
```
Comment (jakule) [inline lib/web/terminal.go:1033]: At that point, we should probably return an error to the client. Without it client potentially will wait for a response that will never come.

File: lib/web/terminal.go [hunk 7]
```diff
@@ -1068,6 +998,199 @@ func (t *WSStream) writeError(err error) error {
...
+				continue
+			}
+
+			var envelope Envelope
+			if err := proto.Unmarshal(bytes, &envelope); err != nil {
+				t.log.Warnf("Unable to parse message payload %v", err)
```
Comment (jakule) [inline lib/web/terminal.go:1039]: The same as above.

File: lib/web/terminal.go [hunk 8]
```diff
@@ -166,6 +170,11 @@ type TerminalHandlerConfig struct {
 	tracer oteltrace.Tracer
 	// ParticipantMode is the mode that determines what you can do when you join an active session.
 	ParticipantMode types.SessionParticipantMode
+	// Tracker is the session tracker of the session being joined. May be nil
+	// if the user is not joining a session.
+	Tracker types.SessionTracker
+	// Clock used for presence checking.
+	Clock clockwork.Clock
```
Comment (jakule) [inline lib/web/terminal.go:177]: Does the clock need to be public?

File: lib/client/client.go [hunk 9]
```diff
@@ -1691,7 +1691,7 @@ func (c *NodeClient) handleGlobalRequests(ctx context.Context, requestCh <-chan
 					continue
 				}
 
-				c.OnMFA()
+				go c.OnMFA()
```
Comment (jakule) [inline lib/client/client.go:1694]: Can you add a comment on why this call is not async?

File: lib/client/presence.go [hunk 10]
```diff
@@ -23,26 +23,59 @@ import (
...
+// PresenceMaintainer allows maintaining presence with the Auth service.
+type PresenceMaintainer interface {
+	MaintainSessionPresence(ctx context.Context) (proto.AuthService_MaintainSessionPresenceClient, error)
+}
+
+const mfaChallengeInterval = time.Second * 30
+
+// PresenceOptions allows passing optional overrides
+// to RunPresenceTask. Mainly used by tests.
+type PresenceOptions struct {
```
Comment (zmb3) [inline lib/client/presence.go:40]: Maybe unexport this to keep the API surface smaller? (Other packages only use the exported `WithXXX` functions and don't need to know about the internal struct)

**Review by zmb3 — APPROVED** [review]

**Pull request merged**